lib/HTTP/Response.pm

```diff
@@ -126,11 +126,30 @@ sub filename
     if ($cd) {
 	require HTTP::Headers::Util;
 	if (my @cd = HTTP::Headers::Util::split_header_words($cd)) {
-	    my ($disposition, undef, %cd_param) = @{$cd[-1]};
-	    $file = $cd_param{filename};
-
-	    # RFC 2047 encoded?
-	    if ($file && $file =~ /^=\?(.+?)\?(.+?)\?(.+)\?=$/) {
+		my ($disposition, undef, %cd_param) = @{$cd[-1]};
+
+		$file = $cd_param{filename};
+
+		if( exists $cd_param{'filename*'} &&
+				$cd_param{'filename*'} =~ m/^([^']+)'(.*)'(.+)$/ ) {
+			# RFC 8187 encoded? Takes precedence.
+			my $charset = $1;
+			my $lang = $2;
+			my $encfile = $3;
+
+			warn q|Producers should use either UTF-8 or ISO-8859-1 character encoding.| unless $charset =~ /^(utf-8|iso-8859-1)$/i;
+
+			local($SIG{__DIE__});
+			eval {
+				require URI::Escape;
+				require Encode;
+				$encfile = URI::Escape::uri_unescape($encfile);
+				Encode::from_to($encfile, $charset, "locale_fs");
+			};
```

Comment:
We are already using Try::Tiny in this module -- that is much preferable to manual exception handling (and altering the die handler should not be necessary).